src/display/edition/freetext.js

```diff
@@ -0,0 +1,225 @@
+/* Copyright 2022 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { AnnotationEditorType, Util } from "../../shared/util.js";
+import { AnnotationEditor } from "./editor.js";
+import { bindEvents } from "./tools.js";
+
+/**
+ * Basic text editor in order to create a FreeTex annotation.
+ */
+class FreeTextEditor extends AnnotationEditor {
+  #color;
+
+  #content = "";
+
+  #contentHTML = "";
+
+  #fontSize;
+
+  static _freeTextDefaultContent = "";
+
+  constructor(params) {
+    super({ ...params, name: "freeTextEditor" });
+    this.#color = params.color || "CanvasText";
+    this.#fontSize = params.fontSize || 10;
+  }
+
+  static setL10n(l10n) {
+    l10n?.get("freetext_default_content").then(msg => {
+      this._freeTextDefaultContent = msg;
+    });
+  }
+
+  /** @inheritdoc */
+  copy() {
+    const editor = new FreeTextEditor({
+      parent: this.parent,
+      id: this.parent.getNextId(),
+      x: this.x,
+      y: this.y,
+    });
+
+    editor.width = this.width;
+    editor.height = this.height;
+    editor.#color = this.#color;
+    editor.#fontSize = this.#fontSize;
+    editor.#content = this.#content;
+    editor.#contentHTML = this.#contentHTML;
+
+    return editor;
+  }
+
+  /** @inheritdoc */
+  rebuild() {
+    if (this.div === null) {
+      return;
+    }
+
+    if (!this.isAttachedToDOM) {
+      // At some point this editor has been removed and
+      // we're rebuilting it, hence we must add it to its
+      // parent.
+      this.parent.add(this);
+    }
+  }
+
+  /** @inheritdoc */
+  enableEditMode() {
+    super.enableEditMode();
+    this.overlayDiv.classList.remove("enabled");
+    this.div.draggable = false;
+  }
+
+  /** @inheritdoc */
+  disableEditMode() {
+    super.disableEditMode();
+    this.overlayDiv.classList.add("enabled");
+    this.div.draggable = true;
+  }
+
+  /** @inheritdoc */
+  onceAdded() {
+    if (this.width) {
+      // The editor has been created in using ctrl+c.
+      this.div.focus();
+      return;
+    }
+    this.enableEditMode();
+    this.editorDiv.focus();
+  }
+
+  /** @inheritdoc */
+  isEmpty() {
+    return this.editorDiv.innerText.trim() === "";
```

Comment:
Here, and elsewhere below, can we perhaps use `textContent` since I believe that's generally recommended over `innerText` these days?

Reply:
Unfortunately, I think `innerText` is better than `textContent` mainly because of `<br>`:
https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/innerText
I wanted to use a basic `textarea` but I didn't manage to make it growable so I fall back on a `div`, hence the `<br>`...